Added OFA support for Apple M4

mmoelle1 opened the pull request:
Added support for Apple M4 chips in optimization-for-architecture

mmoelle1 committed:
Added OFA support for Apple M4

filiatra commented:
There is also a problem on arm:
https://launchpadlibrarian.net/786820763/buildlog_ubuntu-xenial-arm64.gismo_202504061901-f2fe4d3e~ubuntu16.04.1_BUILDING.txt.gz

```
-- (!) aarch64 architecture optimization flags applied: -march=armv8.1-a+nofp+nosimd+nocrc+nocrypto+nolse. In case of runtime errors re-compile with
```

Is there a way to fix it ?
```
/<<PKGBUILDDIR>>/src/gsUtils/gsMesh/gsVertex.h: In member function ‘T gismo::gsVertex<T>::x() const [with T = double]’:
/<<PKGBUILDDIR>>/src/gsUtils/gsMesh/gsVertex.h:120:16: error: ‘+nofp’ feature modifier is incompatible with floating-point return type
```

filiatra commented:
Also, on my new intel machine:
```
CMake Warning at cmake/ofa/AutodetectX86.cmake:290 (message):
   Your CPU is not known.
   Auto-detection of optimization flags failed and will use the 65nm Core 2 CPU settings.
   Please send an email to gismo@inria.fr with the following content so that we can update the OFA script:
   Vendor id:    GenuineIntel
   CPU family:   6
   CPU mode:     170
   CPU stepping: 4
   CPU flags:    vnmi preemption_timer posted_intr invvpid ept_x_only ept_ad ept_1gb flexpriority apicv tsc_offset vtpr mtf vapic ept vpid unrestricted_guest vapic_reg vid ple shadow_vmcs pml ept_mode_based_exec tsc_scaling usr_wait_pause notify_vm_exiting
Call Stack (most recent call first):
  cmake/OptimizeForArchitecture.cmake:80 (OFA_AutodetectX86)
  cmake/OptimizeForArchitecture.cmake:121 (OFA_AutodetectHostArchitecture)
  cmake/gsConfig.cmake:277 (OptimizeForArchitecture)
  CMakeLists.txt:117 (include)
```

mmoelle1 commented:
I will check both issues before merging this PR.

filiatra commented:
I would prefer to have the optimizations displayed it in the options,
and set it to none by default, so that the user may enable it with -D TARGET_ARCHITECTURE=auto
Something like:

```
   GGGGGGGGG      GGGG      GGGGGGGGG  GGGGGG   GGGGGG  GGGGGGGGGG
  GGGG            GGGG     GGGG        GGGGGG  GGGGGG  GGGG   GGGG
 GGGG         GGGGGGGGGGGG GGGGGGGGG   G GGGG  G GGGG GGGG    GGGG
 GGGG GGGGGG GGGGGGGGGGGGG GGGGGGGGGG GG GGGG GG GGGG GGGG   GGGGG
GGGGG  GGGGG GGGGGGGGGGGG  GGGGGGGGG  GG GGGGGG GGGG  GGGG   GGGG 
GGGG   GGGG      GGGG           GGGG  GG  GGGG  GGGG  GGGG   GGGG 
 GGGGGGGGGG      GGGG     GGGGGGGGG  GG   GGG   GGGG  GGGGGGGGGG  
                          version 25.1.0
Configuration (cmake 3.22.1):
  Source folder:          /home/amantzaf/gismo/stable
  Git commit:             0f348cb7
  CMAKE_BUILD_TYPE        Release
  TARGET_ARCHITECTURE     none
  CMAKE_C_COMPILER        /usr/bin/cc
  CMAKE_CXX_COMPILER      /usr/bin/c++
  CMAKE_CXX_STANDARD      14
  GISMO_COEFF_TYPE        double
  GISMO_INDEX_TYPE        int
  GISMO_SHORT_TYPE        int
  GISMO_BUILD_EXAMPLES    ON
  GISMO_BUILD_LIB         ON
-- Zlib found /usr/lib/x86_64-linux-gnu/libz.so
  GISMO_OPTIONAL        
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
  CMAKE_INSTALL_PREFIX    /usr/local
-- Number of examples to compile: 60
-- Configuring done
-- Generating done
```

what do you think ?

mmoelle1 committed:
Optimization for Architecture disabled by default

OFA is set to 'none' by default and needs to be enabled by passing -DTARGET_ARCHITECTURE=auto or a specific architecture during CMake configuration

mmoelle1 committed:
Updated OFA for x86_64 CPUs

mmoelle1 committed:
Missing CPU definition

filiatra committed:
improve script so that the OFA runs (once) when needed

filiatra committed:
set default build to Release

filiatra commented:
Looks good!

On c++ (GCC) 14.1.1 20240620 (Red Hat 14.1.1-6)

```
-- Vendor id:    GenuineIntel
-- CPU family:   6
-- CPU mode:     85
-- CPU stepping: 7
-- Detected Host CPU: cascadelake
-- Checking Host CPU features. This can take some time ...
-- (!) x86_64 architecture optimization flags applied: -march=cascadelake;-mmmx;-mavx;-mavx2;-mfma;-msse2;-msse3;-msse4.1;-msse4.2;-msse;-mssse3;-mavx512bw;-mavx512cd;-mavx512dq;-mavx512f;-mavx512vl;-mavx512vnni;-madx;-maes;-mbmi2;-mf16c;-mfsgsbase;-mfxsr;-mlzcnt;-mpclmul;-mpku;-mpopcnt;-mprfchw;-mrdrnd;-mrdseed;-mxsavec;-mxsaves;-mbmi;-mclflushopt;-mclwb;-mno-sse4a;-mno-avxvnni;-mno-avx5124fmaps;-mno-avx5124vnniw;-mno-avx512bf16;-mno-avx512bitalg;-mno-avx512er;-mno-avx512fp16;-mno-avx512ifma;-mno-avx512pf;-mno-avx512vbmi2;-mno-avx512vbmi;-mno-avx512vp2intersect;-mno-avx512vpopcntdq;-mno-amx-bf16;-mno-amx-int8;-mno-amx-tile;-mno-enqcmd;-mno-gfni;-mno-hreset;-mno-movbe;-mno-movdir64b;-mno-movdiri;-mno-mpx;-mno-pconfig;-mno-prefetchwt1;-mno-ptwrite;-mno-rdpid;-mno-rtm;-mno-serialize;-mno-sha;-mno-tsxldtrk;-mno-uintr;-mno-vaes;-mno-vpclmulqdq;-mno-waitpkg;-mno-wbnoinvd;-mno-xsaveopt;-mno-abm;-mno-cldemote. In case of runtime errors re-compile with
```
there are these 5 warnings appearing on each file compiling:
```
c++: warning: AVX5124FMAPS support will be removed in GCC 15
c++: warning: AVX5124VNNIW support will be removed in GCC 15
c++: warning: AVX512ER support will be removed in GCC 15
c++: warning: AVX512PF support will be removed in GCC 15
c++: warning: PREFETCHWT1 support will be removed in GCC 15
```
Should update something to remove them ? I do not really know the rationale behind these upcoming changes of GCC.
